src/Adapter/Module/Preloader/ModuleServicePreloader.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Module\Preloader;
+
+use PrestaShop\PrestaShop\Core\Addon\Module\ModulePreloaderInterface;
+use PrestaShopBundle\DependencyInjection\RuntimeContainerBuilder;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
+
+/**
+ * Class RuntimeServiceLoader is able to dynamically load the services defined by a module during
+ * runtime, which allows to use them even during installation process (usually only service of installed
+ * modules are loaded).
+ */
+final class ModuleServicePreloader implements ModulePreloaderInterface
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * @var string
+     */
+    private $modulesFolder;
+
+    /**
+     * @param ContainerInterface $container
+     * @param string $modulesFolder
+     */
+    public function __construct(
+        ContainerInterface $container,
+        string $modulesFolder
+    ) {
+        $this->container = $container;
+        $this->modulesFolder = $modulesFolder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preload(string $moduleName): void
+    {
+        // Load module services
+        $serviceConfigPaths = [
+            $this->modulesFolder . '/' . $moduleName . '/config/admin/',
+            $this->modulesFolder . '/' . $moduleName . '/config/',
+        ];
+
+        $runtimeBuilder = new RuntimeContainerBuilder($this->container);
```

Comment:
This instance can be in the constructor, same for $loader. Both are not related to the `$moduleName` and will be created each time you preload a module.

Reply:
Actually it's not that simple, the loader depends on the module because they load the config inside this module (the `$serviceConfigPath` variable depends on the module location)

And I need a new instance of `RuntimeContainerBuilder` each time, because when `$loader->load('services.yml');` is called it actually fills the `$runtimeBuilder` with the service definitions contained in the `services.yml` file

The purpose of this class is to simulate a new container building, parse the services present in the modules, and then, during the compilation process, "copy" them into the current container. So each time `preload` is called a new `RuntimeContainerBuilder` instance is needed (because it can't be compiled twice), but anyway it should be called only once since only one module is installed per process